File: core/tests/lengthTokens.test.cpp
#include "doctest.h"
#include "lengthTokens.hpp"

using namespace stencil::core;

TEST_CASE("parseLengthToken: bare numbers are signed deltas") {
  auto a = parseLengthToken("10");
  REQUIRE(a);
  CHECK(a->kind == LengthKind::Delta);
  CHECK(a->value == doctest::Approx(10.0));

  auto b = parseLengthToken("-10");
  REQUIRE(b);
  CHECK(b->kind == LengthKind::Delta);
  CHECK(b->value == doctest::Approx(-10.0));  // sign folded into the delta

  auto c = parseLengthToken(".5");
  REQUIRE(c);
  CHECK(c->value == doctest::Approx(0.5));
}

TEST_CASE("parseLengthToken: units and the from-end '-' flag") {
  auto px = parseLengthToken("120px");
  REQUIRE(px);
  CHECK(px->kind == LengthKind::Px);
  CHECK(px->value == doctest::Approx(120.0));
  CHECK_FALSE(px->fromEnd);

  auto pct = parseLengthToken("-60%");
  REQUIRE(pct);
  CHECK(pct->kind == LengthKind::Percent);
  CHECK(pct->value == doctest::Approx(60.0));  // magnitude only
  CHECK(pct->fromEnd);                          // '-' means measured from the end

  CHECK(parseLengthToken("3cm")->value == doctest::Approx(3.0));
  CHECK(parseLengthToken("5mm")->value == doctest::Approx(0.5));   // mm -> cm/10
  CHECK(parseLengthToken("1in")->value == doctest::Approx(2.54));  // in -> cm*2.54
  CHECK(parseLengthToken("1in")->kind == LengthKind::Cm);
}

TEST_CASE("parseLengthToken: rejects malformed input and unknown units") {
  CHECK_FALSE(parseLengthToken(""));
  CHECK_FALSE(parseLengthToken("abc"));
  CHECK_FALSE(parseLengthToken("5."));     // trailing dot
  CHECK_FALSE(parseLengthToken("10ln"));   // not a defined unit
  CHECK_FALSE(parseLengthToken("10 p x"));
}

TEST_CASE("resolveAxisPx: absolute and relative resolution") {
  // 50% of a 200px axis = 100px; from-end mirrors it (still 100 here).
  CHECK(*resolveAxisPx("50%", 200, 10, 0) == doctest::Approx(100.0));
  CHECK(*resolveAxisPx("-25%", 200, 10, 0) == doctest::Approx(150.0));
  // cm uses px-per-cm; px is literal.
  CHECK(*resolveAxisPx("2cm", 200, 10, 0) == doctest::Approx(20.0));
  CHECK(*resolveAxisPx("37px", 200, 10, 0) == doctest::Approx(37.0));
  // A bare number is a delta from currentPx.
  CHECK(*resolveAxisPx("5", 200, 10, 100) == doctest::Approx(105.0));
  CHECK_FALSE(resolveAxisPx("nope", 200, 10, 0));
}

File: browser/tests/lengthTokens.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { parseLengthToken, resolveAxisPx, normalizePageSize, isDeltaToken } from '../js/core/units.js';

// Pure length-token parsing used by the console API (window.stencil) for crop/move.

test('parseLengthToken: numbers are px deltas (sign preserved)', () => {
  assert.deepEqual(parseLengthToken(10), { kind: 'delta', value: 10 });
  assert.deepEqual(parseLengthToken(-4), { kind: 'delta', value: -4 });
  assert.equal(parseLengthToken(NaN), null);
  assert.deepEqual(parseLengthToken('12'), { kind: 'delta', value: 12 });
  assert.deepEqual(parseLengthToken('-7'), { kind: 'delta', value: -7 });
});

test('parseLengthToken: units → cm, with from-end flag on leading minus', () => {
  assert.deepEqual(parseLengthToken('3cm'), { kind: 'cm', value: 3, fromEnd: false });
  assert.deepEqual(parseLengthToken('-4cm'), { kind: 'cm', value: 4, fromEnd: true });
  assert.deepEqual(parseLengthToken('1in'), { kind: 'cm', value: 2.54, fromEnd: false });
  assert.deepEqual(parseLengthToken('20mm'), { kind: 'cm', value: 2, fromEnd: false });
  assert.deepEqual(parseLengthToken('50px'), { kind: 'px', value: 50, fromEnd: false });
});

test('parseLengthToken: percent + bad input', () => {
  assert.deepEqual(parseLengthToken('50%'), { kind: 'percent', value: 50, fromEnd: false });
  assert.deepEqual(parseLengthToken('-60%'), { kind: 'percent', value: 60, fromEnd: true });
  assert.equal(parseLengthToken('abc'), null);
  assert.equal(parseLengthToken(''), null);
  assert.equal(parseLengthToken(null), null);
});

test('resolveAxisPx: deltas add to current; absolutes from start/end', () => {
  const ctx = { lengthPx: 1000, pxPerCm: 10, currentPx: 100 };
  assert.equal(resolveAxisPx(25, ctx), 125);              // +25 px from current
  assert.equal(resolveAxisPx('200px', ctx), 200);         // absolute px
  assert.equal(resolveAxisPx('5cm', ctx), 50);            // 5cm * 10 px/cm
  assert.equal(resolveAxisPx('-5cm', ctx), 950);          // from end: 1000 - 50
  assert.equal(resolveAxisPx('50%', ctx), 500);           // half the axis
  assert.equal(resolveAxisPx('-60%', ctx), 400);          // 1000 - 600
  assert.equal(resolveAxisPx('nope', ctx), null);
});

test('normalizePageSize: case-insensitive over the whole ISO table, else null', () => {
  assert.equal(normalizePageSize('a3'), 'A3');
  assert.equal(normalizePageSize('A4'), 'A4');
  assert.equal(normalizePageSize('B5'), 'B5');
  assert.equal(normalizePageSize('a10'), 'A10');
  assert.equal(normalizePageSize('c10'), 'C10');
  assert.equal(normalizePageSize('Custom'), 'custom');
  assert.equal(normalizePageSize('D4'), null);
  assert.equal(normalizePageSize(''), null);
});

test('isDeltaToken: only bare numbers are deltas', () => {
  assert.equal(isDeltaToken(5), true);
  assert.equal(isDeltaToken('5'), true);
  assert.equal(isDeltaToken('5cm'), false);
  assert.equal(isDeltaToken('50%'), false);
});
